### PR TITLE
Refactor KBreadcrumbs to utilize KListWithOverflow

### PR DESCRIPTION
**Pull request by sruthin21**
## Description

Task1 
 1. Added OverFlowDirection Prop for the kListwithoverflow component
 Task2
 1. Updated KBreadcrums to use KListWithOverflow.

#### Issue addressed

Addresses #693.

## Changelog

  - **Description:** Summary of change(s)
  - **Products impact:** Choose from - none (for internal updates) / bugfix / new API / updated API / removed API. If it's 'none', use "-" for all items below to indicate they are not relevant.
  - **Addresses:** Link(s) to GH issue(s) addressed. Include KDS links as well as links to related issues in a consumer product repository too.
  - **Components:** Affected public KDS component. Do not include internal sub-components or documentation components.
  - **Breaking:** Will this change break something in a consumer? Choose from: yes / no
  - **Impacts a11y:** Does this change improve a11y or adds new features that can be used to improve it? Choose from: yes / no
  - **Guidance:** Why and how to introduce this update to a consumer? Required for breaking changes, appreciated for changes with a11y impact, and welcomed for non-breaking changes when relevant.

## Steps to test

1. Step 1
2. Step 2
3. ...

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [ ] The change is described in the changelog section above

## Reviewer guidance

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

### CODE REVIEW

File: lib/KBreadcrumbs.vue [hunk 1]
```diff
@@ -1,378 +1,110 @@
...
-      showSingleItem: {
-        type: Boolean,
-        default: false,
-      },
+import KListWithOverflow from './KListWithOverflow.vue';
+import KDropdownMenu from './KDropdownMenu.vue';
+export default {
+  name: 'KBreadcrumbs',
+  props: {
+    breadcrumbs: {
```
Comment (AlexVelezLl) [inline lib/KBreadcrumbs.vue:45]: Here we should not change the public API as this would be a breaking change, we should keep the same API as before, or at least that if we make any change to the API, that this is not a breaking change. So we should keep the `items` and `showSingleItem` props, and they should keep working exactly the same as before the refactor.

File: lib/KBreadcrumbs.vue [hunk 2]
```diff
@@ -1,378 +1,110 @@
...
-      },
-      /**
-       * By default, the breadcrums will be hidden when the length of items is 1.
-       * When set to `true`, a breadcrumb will be shown even when there is only one.
-       */
-      showSingleItem: {
-        type: Boolean,
-        default: false,
-      },
+import KListWithOverflow from './KListWithOverflow.vue';
```
Comment (AlexVelezLl) [inline lib/KBreadcrumbs.vue:40]: Here we dont need to import there components, as they are already registered in the Vue context, we can use it without it.

File: lib/KListWithOverflow.vue [hunk 3]
```diff
@@ -195,11 +207,17 @@
...
 
-        maxWidth = Math.ceil(maxWidth);
-        this.overflowItems = overflowItemsIdx.map(idx => this.items[idx]);
+        this.overflowItems = overflowItemsIdx.map((idx) => this.items[idx]);
         this.isMoreButtonVisible = overflowItemsIdx.length > 0;
-        list.style.maxWidth = `${maxWidth}px`;
+        list.style.maxWidth = `${Math.ceil(maxWidth)}px`;
         list.style.maxHeight = `${maxHeight}px`;
+
+        this.setVisibleItems(directionIndexes, overflowItemsIdx);
```
Comment (AlexVelezLl) [inline lib/KListWithOverflow.vue:215]: We are not doing anything with this `visibleItems` array. I would suggest to remove them. Also I dont note any added value of removing the `maxWidth = Math.ceil(maxWidth);` sentence. I think we can keep these lines as before:
```javascript
        maxWidth = Math.ceil(maxWidth);
        this.overflowItems = overflowItemsIdx.map(idx => this.items[idx]);
        this.isMoreButtonVisible = overflowItemsIdx.length > 0;
        list.style.maxWidth = `${maxWidth}px`;
        list.style.maxHeight = `${maxHeight}px`;
      },
```

File: lib/KListWithOverflow.vue [hunk 4]
```diff
@@ -152,29 +161,32 @@
           itemsSizes.push(itemSize);
         }
 
+        if (this.overflowDirection === 'start') {
+              itemsSizes.reverse();
```
Comment (AlexVelezLl) [inline lib/KListWithOverflow.vue:165]: We dont need to reverse the itemSizes because we are already reverting the `directionIndexes`, and when we iterate `directionIndexes`, we are using the same index to access the list.children which is not reverted and the itemSizes which is reverted. So we are not matching the same information for the children and its size.

File: lib/KListWithOverflow.vue [hunk 5]
```diff
@@ -152,29 +161,32 @@
...
-          // If the item dont fit in the available space or if we have already
-          // overflowed items, we hide it. This means that once one item overflows,
-          // all the following items will be hidden.
+        directionIndexes.forEach((i) => {
+          const itemWidth = itemsSizes[i].width;
           if (itemWidth >= availableWidth || overflowItemsIdx.length > 0) {
             overflowItemsIdx.push(i);
-            item.style.visibility = 'hidden';
-            item.style.position = 'absolute';
+            list.children[i].style.visibility = 'hidden';
```
Comment (AlexVelezLl) [inline lib/KListWithOverflow.vue:177]: I think its better to revert this change and have the `item` object to avoid repeating the `list.children[i]`

File: lib/KListWithOverflow.vue [hunk 6]
```diff
@@ -152,29 +161,32 @@
...
+        const indexSequence = [...Array(list.children.length).keys()];
+        const directionIndexes =
+          this.overflowDirection === 'start' ? indexSequence.reverse() : indexSequence;
+
         const overflowItemsIdx = [];
-        for (let i = 0; i < list.children.length; i++) {
-          const item = list.children[i];
-          const itemWidth = itemsSizes[i].width;
 
-          // If the item dont fit in the available space or if we have already
```
Comment (AlexVelezLl) [inline lib/KListWithOverflow.vue:160]: Lets keep these comments, as they are useful for new team members that doesnt know the logic here

File: lib/KListWithOverflow.vue [hunk 7]
```diff
@@ -195,11 +207,17 @@
           maxWidth -= removedDividerWidth;
```
Comment (AlexVelezLl) [inline lib/KListWithOverflow.vue:207]: We need to also check the logic of the fix dividers Visibility, because for example here, when we have dividers, we can have this situation where some overflowed items are visible:

https://github.com/user-attachments/assets/ac572531-7806-4462-a841-45a08f04bc37

File: lib/KListWithOverflow.vue [hunk 8]
```diff
@@ -276,5 +294,8 @@
   .more-button-wrapper {
     visibility: hidden;
   }
+  .more-button-wrapper.start-button {
```
Comment (AlexVelezLl) [inline lib/KListWithOverflow.vue:297]: We cant have this solution for the position of the button, because Keyboard navigation would not be the same. We need to keep the current behaviour in KBreadcrumbs that if the more button appears at the beginning, the keyboard navigation should focus it first, before the rest items.

https://github.com/user-attachments/assets/dbf2abb9-4a9e-471b-be0f-eb2081e959f3

File: lib/KBreadcrumbs.vue [hunk 9]
```diff
@@ -1,378 +1,110 @@
...
+  margin-top: 8px;
+  margin-bottom: 8px;
+  font-size: 16px;
+  font-weight: bold;
+  line-height: 32px;
+  white-space: nowrap;
+}
+
+.link {
+  color: blue;
```
Comment (AlexVelezLl) [inline lib/KBreadcrumbs.vue:82]: Here we shouldnt add color styles like these. We cant set burned colors, instead we need to use our brand colors, you can find the guides [here](https://design-system.learningequality.org/colors/).

Although, I think most of these styles arent really needed. Do they?

**Review by AlexVelezLl — CHANGES_REQUESTED** [review]
Hey @sruthin21! Thanks for your hard work. This is looking great! However my major concern is that we should make this change a non-breaking change, and avoid modifying the public API of the KBreadcrumbs and keep the external functionality the same as before.

One method to ensure that we are not breaking anything is to constantly check the KBreadcrumbs page http://localhost:4000/kbreadcrumbs and see that it looks and behave exactly the same as our current version of kbreadcrumbs which you can see in the production docs site https://design-system.learningequality.org/kbreadcrumbs/.

As you can see now we dont show the breadcrumbs in the docs page:

| Before | After |
| --- | --- |
| image | image |

**Comment by AlexVelezLl** [comment]
Also, a side note! When we open a PR we should keep and fill the form template that appears when we create the PR, not remove it, I have pasted it again to the PR description :hugs:.

**Comment by AlexVelezLl** [comment]
Hey @sruthin21! Are you still interested in continuing working in this issue?

**Comment by MisRob** [comment]
Hi @sruthin21, by deleting your fork, unfortunately we won't be able to finish this PR. I will unassign you from https://github.com/learningequality/kolibri-design-system/issues/693, but feel free to message us and re-open if you'd like.
